### PR TITLE
Issue #64 - Pytorch Tutorial 3 - Jupter Notebook and Markdown

### PR DESCRIPTION
**Pull request by Adamage**
Various comments and markdown added to the source script, inspired by the "old" readme.md file.

Closing #64

### CODE REVIEW

File: tutorials/pytorch/tut3_mixed_precision/README.md [hunk 1]
```diff
@@ -110,43 +109,81 @@ class CustomModel(nn.Module):
...
-                                transforms.ConvertImageDtype(torch.half)])
+if data_half:
+    transform = transforms.Compose([transforms.Resize(128),
+                                    transforms.ToTensor(),
+                                    transforms.Normalize((0.5,), (0.5,)),
+                                    transforms.ConvertImageDtype(torch.half)])
+else:
+    transform = transforms.Compose([transforms.Resize(128),
+                                    transforms.ToTensor(),
+                                    transforms.Normalize((0.5,), (0.5,))])
```
Comment (plutasnyy) [inline tutorials/pytorch/tut3_mixed_precision/README.md:186]: It could also be written as:
```python
transform_list = [transforms.Resize(128),
                  transforms.ToTensor(),
                  transforms.Normalize((0.5,), (0.5,)]
if data_half:
    transform_list += transforms.ConvertImageDtype(torch.half)

transform = transforms.Compose(transform_list)
```

File: tutorials/pytorch/tut3_mixed_precision/README.md [hunk 2]
```diff
@@ -158,56 +195,93 @@ test_dataset = torchvision.datasets.FashionMNIST("./datasets/",
...
-                                 accum_type=torch.float16)
+if optimizer_half:
+    optimizer = poptorch.optim.AdamW(model.parameters(),
+                                     lr=0.001,
+                                     loss_scaling=1024,
+                                     accum_type=torch.float16)
+else:
+    optimizer = poptorch.optim.AdamW(model.parameters(),
+                                     lr=0.001,
+                                     accum_type=torch.float32)
```
Comment (plutasnyy) [inline tutorials/pytorch/tut3_mixed_precision/README.md:223]: It could also be written as:
```python
accum = torch.float16 if optimzer_half else torch.float32
loss_scaling = 1024 if optimizer_half else default_value
optimizer = poptorch.optim.AdamW(model.parameters), lr=0.001, accum_type=torch.float32)
```

File: tutorials/pytorch/tut3_mixed_precision/README.md [hunk 3]
```diff
@@ -219,72 +293,105 @@ train_dataloader = poptorch.DataLoader(opts,
 
 We first make sure our model is in training mode, and then wrap it with `poptorch.trainingModel`.
 
+
 ```python
-model.train()
```
Comment (plutasnyy) [inline tutorials/pytorch/tut3_mixed_precision/README.md:223]: Wasn't it needed?

Comment (Adamage) [reply tutorials/pytorch/tut3_mixed_precision/README.md:223]: So actually this one is tricky, nice that you spotted this. When you initialize the model it's already in the `train` mode. Or in other words `model.train` is `True`. Only after you run `model.eval()` it switches back to `False` and you need `model.train()` to switch it back. 

I actually removed this by accident, but it doesn't do anything in this specific place and can be removed. Maybe for clarity it can be left. Let's see

File: tutorials/pytorch/tut3_mixed_precision/walkthrough.py [hunk 4]
```diff
@@ -83,9 +188,25 @@ def forward(self, x, labels=None):
                                                  transform=transform,
```
Comment (Adamage) [inline tutorials/pytorch/tut3_mixed_precision/walkthrough.py:188]: FashionMNIST - split to another cell with description above

File: tutorials/pytorch/tut3_mixed_precision/README.md (hunk 2 above)
Comment (Adamage) [reply tutorials/pytorch/tut3_mixed_precision/README.md]: done

File: tutorials/pytorch/tut3_mixed_precision/README.md (hunk 1 above)
Comment (Adamage) [reply tutorials/pytorch/tut3_mixed_precision/README.md]: done

File: tutorials/pytorch/tut3_mixed_precision/walkthrough.py (hunk 4 above)
Comment (Adamage) [reply tutorials/pytorch/tut3_mixed_precision/walkthrough.py:188]: done

**Comment by Adamage** [comment]
- [ ] Apply limit of `79 chars` per line (except for web links)
- [ ] regenerate when #50  is updated

File: tutorials/pytorch/tut3_mixed_precision/README.md [hunk 5]
```diff
@@ -110,44 +127,94 @@ class CustomModel(nn.Module):
...
+transform_list = [transforms.Resize(128),
+                  transforms.ToTensor(),
+                  transforms.Normalize((0.5,), (0.5,))]
+if data_half:
+    transform_list.append(transforms.ConvertImageDtype(torch.half))
+
+transform = transforms.Compose(transform_list)
+```
+
+Pull the datasets if they are not available locally:
```
Comment (plutasnyy) [inline tutorials/pytorch/tut3_mixed_precision/README.md:214]: I think trochvision downloads the dataset itself ;P

Comment (Adamage) [reply tutorials/pytorch/tut3_mixed_precision/README.md:214]: Yes but if it's the second time you run it, it does not.

File: tutorials/pytorch/tut3_mixed_precision/README.md [hunk 6]
```diff
@@ -158,76 +225,125 @@ test_dataset = torchvision.datasets.FashionMNIST("./datasets/",
...
+optimisers in `poptorch.optim`. In this tutorial, we will set it to 1024 for 
+an AdamW optimizer. For all optimisers (except `poptorch.optim.SGD`), 
+using a model in FP16 requires the argument `accum_type` to be set to 
+`torch.float16` as well:
+
 
 ```python
-optimizer = poptorch.optim.AdamW(model.parameters(),
+accum, loss_scaling = \
+    (torch.float16, 1024) if optimizer_half else (torch.float32, None)
```
[the remark this answers is not in the document]
Comment (Adamage) [reply tutorials/pytorch/tut3_mixed_precision/README.md:249]: Originallyt with args parsing it was:
```python
# Optimizer and loss scaling
if args.optimizer_half:
    optimizer = poptorch.optim.AdamW(model.parameters(),
                                     lr=0.001,
                                     loss_scaling=1024,
                                     accum_type=torch.float16)
else:
    optimizer = poptorch.optim.AdamW(model.parameters(),
                                     lr=0.001,
                                     accum_type=torch.float32)
```

So if `optimizer_half` was `False`, it was not even given to the `AdamW` initializer. And then, the default value in that `__init__` is `None`

**Comment by Adamage** [comment]
@dirkbrink  applied all the discussed changes, we have 2 more open ones, please take a look at my replies, thanks!

File: tutorials/pytorch/tut3_mixed_precision/README.md [hunk 7]
```diff
@@ -237,54 +353,107 @@ for epoch in tqdm(range(epochs), desc="epochs"):
...
+```
+
+Run inference on the labelled data
 
+
+```python
 predictions, labels = [], []
 for data, label in test_dataloader:
-    predictions += poptorch_model_inf(data).data.max(dim=1).indices
+    predictions += poptorch_model_inf(data).data.float().max(dim=1).indices
```
[the remark this answers is not in the document]
Comment (dirkbrink) [reply tutorials/pytorch/tut3_mixed_precision/README.md:387]: Ah I see. Reading the code again, I don't think it is necessarily required, but it forces the predictions to be floats. This all happens after the model has executed so I'm fine to leave this as is, thanks

File: tutorials/pytorch/tut3_mixed_precision/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (dirkbrink) [reply tutorials/pytorch/tut3_mixed_precision/README.md:247]: Alright cool, glad you're spotting these things when I'm not! I don't know why the original author decided to do this, but let's leave it as it is

File: tutorials/pytorch/tut3_mixed_precision/README.md [hunk 8]
```diff
@@ -353,11 +352,12 @@ for epoch in tqdm(range(epochs), desc="epochs"):
         total_loss += loss
 ```
 
-Release IPU resources:
+Release resources:
 
 
 ```python
 poptorch_model.detachFromDevice()
+train_dataloader.terminate()
```
Comment (dirkbrink) [inline tutorials/pytorch/tut3_mixed_precision/README.md:360]: Worth adding in a brief explanation from the Zendesk ticket explaining why the dataloader termination is required (and that it's only required in the Jupyter notebook, but not technically required in a normal Python script)?

Comment (Adamage) [reply tutorials/pytorch/tut3_mixed_precision/README.md:368]: Added just now.

**Review by dirkbrink — CHANGES_REQUESTED** [review]
```
poptorch.poptorch_core.Error: 'poptorch_py_error': AsynchronousDataAccessor worker thread failed to start (Check above for details)
```

This error occurs when trying to run the source and code_only files

**Comment by Adamage** [comment]
@dirkbrink  I see, this happened because yesterday I changed the DataLoader mode to `Async`. In that case, the source script must be guarded by `__main__`. In this tutorial this was not the highlighted functionality, I will revert this mode change.

It is highlighted in other tutorials

File: tutorials/pytorch/tut3_mixed_precision/README.md [hunk 9]
```diff
@@ -158,56 +223,103 @@ test_dataset = torchvision.datasets.FashionMNIST("./datasets/",
...
+The optimal value depends on the model and the training job. This is therefore 
+a hyperparameter for you to tune.
 
 ## Set PopTorch's options
 
-To configure some features of the IPU and to be able to use PopTorch's classes in the next sections, we will need to create an instance of `poptorch.Options` which stores the options we will be using. We covered some of the available options in the [introductory tutorial for PopTorch](https://github.com/graphcore/examples/tree/master/tutorials/pytorch/tut1_basics).
+To configure some features of the IPU and to be able to use PopTorch's classes 
+in the next sections, we will need to create an instance of `poptorch.Options` 
+which stores the options we will be using. We covered some of the available 
+options in the: [Introductory Tutorial for PopTorch](https://github.com/graphcore/examples/tree/master/tutorials/pytorch/tut1_basics).
```
Comment (dirkbrink) [inline tutorials/pytorch/tut3_mixed_precision/README.md:265]: Change back to "in the [introductory tutorial for PopTorch]"
